lib/src/core/thing_discovery.dart

```diff
@@ -86,42 +102,90 @@ class ThingDiscovery extends Stream<ThingDescription>
+    final value = await _servient.contentSerdes.contentToValue(content, null);
+    if (value is CoapWebLink) {
+      return [value];
+    } else if (value is List<CoapWebLink>) {
+      return value;
+    }
+
+    return null;
+  }
+
+  Future<Iterable<Uri>> _filterCoreWebLinks(
+    String resourceType,
+    Content coreWebLink,
+    Uri baseUri,
+  ) async {
+    final webLinks = await _getCorewebLinks(coreWebLink);
+
+    if (webLinks == null) {
+      throw DiscoveryException(
+        'Discovery from $baseUri returned no valid CoRE Link-Format Links.',
+      );
+    }
+
+    return webLinks
+        .where(
+          (element) =>
+              element.attributes.getResourceTypes()?.contains(resourceType) ??
+              false,
+        )
+        .map((weblink) => Uri.tryParse(weblink.uri))
+        .whereType<Uri>()
+        .map((uri) => uri.toAbsoluteUri(baseUri));
+  }
+
   Stream<ThingDescription> _discoverWithCoreLinkFormat(Uri uri) async* {
-    final Set<Uri> discoveredUris = {};
-    await for (final coreWebLink in _client.discoverWithCoreLinkFormat(uri)) {
-      final value =
-          await _servient.contentSerdes.contentToValue(coreWebLink, null);
+    // TODO: Remove additional quotes once fixed in CoAP library
+    yield* _performCoreLinkFormatDiscovery('"wot.thing"', uri)
+        .map(_discoverDirectly)
+        .flatten();
+  }
 
-      if (value is! CoapWebLink ||
-          !(value.attributes.getContentTypes()?.contains('wot.thing') ??
-              false)) {
-        continue;
-      }
+  Stream<ThingDescription> _discoverfromCoreResourceDirectory(Uri uri) async* {
+    // TODO: Remove additional quotes once fixed in CoAP library
+    yield* _performCoreLinkFormatDiscovery('"core.rd-lookup-res"', uri)
+        .map(_discoverWithCoreLinkFormat)
+        .flatten();
+  }
+
+  Stream<Uri> _performCoreLinkFormatDiscovery(
+    String resourceType,
+    Uri uri,
+  ) async* {
+    final Set<Uri> discoveredUris = {};
+    final discoveryUri = uri.toLinkFormatDiscoveryUri(resourceType);
+    final client = _clientForUriScheme(uri);
 
-      Uri? parsedUri = Uri.tryParse(value.uri);
+    await for (final coreWebLink
+        in client.discoverWithCoreLinkFormat(discoveryUri)) {
+      final Iterable<Uri> parsedUris;
 
-      if (parsedUri == null) {
-        // TODO: Should an error be passed on here instead?
+      try {
+        parsedUris =
+            await _filterCoreWebLinks(resourceType, coreWebLink, discoveryUri);
+      } on Exception catch (exception) {
+        yield* Stream.error(exception);
         continue;
       }
 
-      if (!parsedUri.isAbsolute) {
-        parsedUri = parsedUri.replace(
-          scheme: uri.scheme,
-          host: uri.host,
-          port: uri.port,
-        );
-      }
-      if (discoveredUris.contains(parsedUri)) {
-        continue;
+      for (final parsedUri in parsedUris) {
+        if (discoveredUris.contains(parsedUri)) {
+          continue;
+        }
```

Comment:
Is this check really necessary? `discoveredUris` is already a `Set`, so adding an element which is already present shouldn't have any effect.

Reply:
Hmm, very good point regarding the `Set`. I refactored the code in 83d6125a185e530334d28b7bf3e4944d16331934, taking advantage of the boolean return value of the `add` method and making it a bit more concise. Hope that this is an improvement :)

Reply:
Yes, that seems like a good solution 👍🏻